ProjFS.Mac/PrjFSKextTests/HandleOperationTests.mm

```diff
@@ -0,0 +1,569 @@
+            nullptr));
+}
+
+- (void) testDeleteDir {
+    testDirVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_DELETE,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testDirVnode.get()),
+        0,
+        0);
+    XCTAssertTrue(
+       MockCalls::DidCallFunction(
+            ProviderMessaging_TrySendRequestAndWaitForResponse,
+            _,
+            MessageType_KtoU_NotifyDirectoryPreDelete,
+            testDirVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            nullptr));
+    XCTAssertTrue(
+       MockCalls::DidCallFunction(
+            ProviderMessaging_TrySendRequestAndWaitForResponse,
+            _,
+            MessageType_KtoU_RecursivelyEnumerateDirectory,
+            testDirVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            nullptr));
+}
+
+- (void) testEmptyDirectoryEnumerates {
+    testDirVnode->attrValues.va_flags = FileFlags_IsEmpty | FileFlags_IsInVirtualizationRoot;
+    const int actionCount = 5;
+    kauth_action_t actions[actionCount] =
+    {
+        KAUTH_VNODE_LIST_DIRECTORY,
+        KAUTH_VNODE_SEARCH,
+        KAUTH_VNODE_READ_SECURITY,
+        KAUTH_VNODE_READ_ATTRIBUTES,
+        KAUTH_VNODE_READ_EXTATTRIBUTES
+    };
+    
+    for (int i = 0; i < actionCount; i++)
+    {
+        HandleVnodeOperation(
+            nullptr,
+            nullptr,
+            actions[i],
+            reinterpret_cast<uintptr_t>(context),
+            reinterpret_cast<uintptr_t>(testDirVnode.get()),
+            0,
+            0);
+        XCTAssertTrue(
+            MockCalls::DidCallFunction(
+                ProviderMessaging_TrySendRequestAndWaitForResponse,
+                _,
+                MessageType_KtoU_EnumerateDirectory,
+                testDirVnode.get(),
+                _,
+                _,
+                _,
+                _,
+                _,
+                nullptr));
+        MockCalls::Clear();
+    }
+}
+
+- (void) testNonEmptyDirectoryDoesNotEnumerate {
+    testDirVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    const int actionCount = 5;
+    kauth_action_t actions[actionCount] =
+    {
+        KAUTH_VNODE_LIST_DIRECTORY,
+        KAUTH_VNODE_SEARCH,
+        KAUTH_VNODE_READ_SECURITY,
+        KAUTH_VNODE_READ_ATTRIBUTES,
+        KAUTH_VNODE_READ_EXTATTRIBUTES
+    };
+    
+    for (int i = 0; i < actionCount; i++)
+    {
+        HandleVnodeOperation(
+            nullptr,
+            nullptr,
+            actions[i],
+            reinterpret_cast<uintptr_t>(context),
+            reinterpret_cast<uintptr_t>(testDirVnode.get()),
+            0,
+            0);
+        XCTAssertFalse(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+    }
+}
+    
+-(void) testWriteFile {
```

Comment:
Minor naming suggestion:

If the some of the more generic test names were slightly more detailed it might make it easier to find test(s) of interest when looking for specific coverage:

_Examples_

- `testWriteFile ` -> `testWriteDataVnodeOpEmptyFile` 
- `testWriteFileFull` -> `testWriteDataVnodeOpFullFile`

What got me thinking about this was when I read the implementation of `testWriteFileFull` I was surprised to see that it wasn't checking that the `MessageType_KtoU_NotifyFileModified` message was sent.  After realizing that these first tests were only for vnode operations it made sense, but it got me wondering if the test names should include VnodeOp or FileOp (or perhaps the tests should be in different classes?).

Reply:
This is a good point.  I haven't fully decided on seperate files or the same files for both types.  My plan is to revisit decision when I go full force on FileOp tests (seperate PR).  Hence for now I'd like to punt the naming to a future (but not distance future PR).